beacon-chain/db/kv/slasher.go

```diff
@@ -516,11 +500,32 @@ func decodeProposalRecord(encoded []byte) (*slashertypes.SignedBlockHeaderWrappe
 	}
 	signingRoot := encoded[:32]
 	decodedBlkHdr := &ethpb.SignedBeaconBlockHeader{}
-	if err := decodedBlkHdr.UnmarshalSSZ(encoded[32:]); err != nil {
+	decodedHdrBytes, err := snappy.Decode(nil, encoded[32:])
+	if err != nil {
+		return nil, err
+	}
+	if err := decodedBlkHdr.UnmarshalSSZ(decodedHdrBytes); err != nil {
 		return nil, err
 	}
 	return &slashertypes.SignedBlockHeaderWrapper{
 		SignedBeaconBlockHeader: decodedBlkHdr,
 		SigningRoot:             bytesutil.ToBytes32(signingRoot),
 	}, nil
 }
+
+// Encodes an epoch by performing modulo HISTORY_SIZE from slasher using 2 bytes instead of 8 as a
+// client optimization to save space in the database.
+func encodeTargetEpoch(epoch types.Epoch, historyLength uint64) []byte {
+	buf := make([]byte, 2)
+	binary.LittleEndian.PutUint16(buf, uint16(epoch.Mod(historyLength)))
+	return buf
+}
+
+// Encodes a validator index using 5 bytes instead of 8 as a
+// client optimization to save space in the database. Because the max validator
+// registry size is 2**40, this is a safe optimization.
+func encodeValidatorIndex(index types.ValidatorIndex) []byte {
+	buf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(buf, uint64(index))
+	return buf[:5]
```

Comment:
Probably a crazy idea:

Instead of allocating 8 bytes, and then truncating to 5, we can also implement what `PutUint64()` does internally (it is pretty simple function), i.e.:
```golang
func encodeValidatorIndex(index types.ValidatorIndex) []byte {
        buf := make([]byte, 5)
        v := uint64(index)
	buf[0] = byte(v)
	buf[1] = byte(v >> 8)
	buf[2] = byte(v >> 16)
	buf[3] = byte(v >> 24)
	buf[4] = byte(v >> 32)
        return buf
}
```
This is a micro-optimization, of course. But depending on how intense `encodeValidatorIndex` is used, it might be worth considering it.

Reply:
Thanks! Adding